### PR TITLE
Constructing pre-conditioning matrix adaptor with only metric type

### PR DESCRIPTION
**Pull request by yebai**
This PR adds support for constructing pre-conditioning matrix adaptor with only metric type (i.e. diag, unit and dense). 

```julia
pc1 = Preconditioner(:UnitEuclideanMetric) # default dim = 2
```

The adaptor's internal Welford estimator dimension is increased when the first sample is added. 

```julia
# Var adaptor dimention should be increased to length(θ) from 2
θ = [0.0, 0.0, 0.0, 0.0]
AdvancedHMC.adapt!(pc1, θ, 1.)
```
This is required for resolving some adaptor simplification in https://github.com/TuringLang/Turing.jl/pull/739

### CODE REVIEW

File: src/adaptation/precond.jl [hunk 1]
```diff
@@ -111,72 +117,185 @@ function get_cov(wc::WelfordCov{T})::Matrix{T} where {T<:AbstractFloat}
...
 
-reset!(dpc::DensePreConditioner) = reset!(dpc.ce)
+reset!(dpc::DensePreconditioner) = reset!(dpc.ce)
 
-function getM⁻¹(dpc::DensePreConditioner)
+function getM⁻¹(dpc::DensePreconditioner)
     return dpc.covar
 end
+
+####
```
Comment (xukai92) [inline src/adaptation/precond.jl]: Why do we move the code here? Does it make AHMC dependent on Adaptation then?

Comment (xukai92) [reply src/adaptation/precond.jl]: I think ideally we would like AHMC without the Adaptation model will work as well.

File: src/adaptation/precond.jl [hunk 2]
```diff
@@ -111,72 +117,185 @@ function get_cov(wc::WelfordCov{T})::Matrix{T} where {T<:AbstractFloat}
...
+
+function Preconditioner(m::DiagEuclideanMetric)
+    return DiagPreconditioner(m.dim)
+end
+
+function Preconditioner(m::DenseEuclideanMetric)
+    return DensePreconditioner(m.dim)
+end
+
+function Preconditioner(m::Type, dim::Integer=2)
```
Comment (xukai92) [inline src/adaptation/precond.jl]: Why the default is 2? If we want to update is anyway can we just set is as 0 or something

Comment (yebai) [reply src/adaptation/precond.jl]: yeah, it's arbitrary; `0` might work, or might cause an issue in a co-variance matrix.

File: src/adaptation/precond.jl (hunk 1 above)
Comment (yebai) [reply src/adaptation/precond.jl]: I think we can replace `Metric` types with `PDMat` (see https://github.com/TuringLang/AdvancedHMC.jl/issues/34), so this doesn't matter in the future.

Comment (yebai) [reply src/adaptation/precond.jl]: Also, AHMC depends on Adaption anyway: some core features like step-size adaption is impossible without Adaption.

Comment (xukai92) [reply src/adaptation/precond.jl]: But shouldn't a call of `sample()` without adaptation is expected to work?

Comment (yebai) [reply src/adaptation/precond.jl]: I see your point - but I don't think this matters a lot.  Besides, so this dependency will go away if we fix #34, right?

Comment (xukai92) [reply src/adaptation/precond.jl]: Yes that's right! I'm fine with the current code then.

**Comment by yebai** [comment]
I'm still working on the `Turing` side of this PR. Please do not merge yet.

**Comment by yebai** [comment]
@xukai92 This is ready for merge from my side.

File: src/adaptation/stan_adaption.jl [hunk 3]
```diff
@@ -16,14 +16,14 @@ end
 # Acknowledgement: this adaption settings is mimicing Stan's 3-phase adaptation.
 struct StanNUTSAdaptor <: AbstractCompositeAdaptor
     n_adapts    :: Int
-    pc          :: AbstractPreConditioner
+    pc          :: AbstractPreconditioner
```
Comment (mohamed82008) [inline src/adaptation/stan_adaption.jl]: Make a type parameter?

**Pull request merged**